docs: add design document (#1)
vasco-santos/hash-stream-singularity-like-index-store: DESIGN.md

  "carOffset": 2154851,
  "carBlockLength": 197,
  "varint": "wwE=",
  "rawBlock": "EiwKJAFVEiD3fXHsfAtzuy4Rs7Bao4IZFk1sp6n7TiW9l6c8Fsd4ZBIAGICAQBIsCiQBVRIgknipVUA6k+D7/at6kJIKw3djt7YFeTtVtexkvTN3gc8SABiAgEASLAokAVUSIGLF/7OZ+6M6WJKEtwumIV6ucqv+G0o2Z3ZQ1CCgJ1ekEgAYs8EDChMIAhizwYMBIICAQCCAgEAgs8ED",
  "fileOffset": 0,
  "carId": 17482,
  "fileId": null
},
```

That can be transformed to a compatible **IndexRecord** format from HashStream:

```json
{
	"multihash": "MH(bafybeidxxkuao2zamg5rd7pypqrhrjmaqayxp7wr5ojmqdqbtpvzje74au)",
  // Type of the record - Inline Blob
  "type": 4,
  // identity multihash with inline raw block content
  "location": "MH(EiwKJAFVEiD3fXHsfAtzuy4Rs7Bao4IZFk1sp6n7TiW9l6c8Fsd4ZBIAGICAQBIsCiQBVRIgknipVUA6k+D7/at6kJIKw3djt7YFeTtVtexkvTN3gc8SABiAgEASLAokAVUSIGLF/7OZ+6M6WJKEtwumIV6ucqv+G0o2Z3ZQ1CCgJ1ekEgAYs8EDChMIAhizwYMBIICAQCCAgEAgs8ED)",
	// length of the data
	"length": 159,
	// offset of the data in the location byte stream
  "offset": 0,
	// associated records
	"subRecords": []
}
```

The `rawBlock` content should be the byte representation that hashes to the given CID. Therefore, its content can be streamed back by HashStream to the client.

> Once the client verifies and decodes the block, it MAY intend to traverse the DAG and request following blocks from the server. For instance, a follow-up request MAY be performed with CID `bafkreihxpvy6y7aloo5s4entwbnkhaqzczgwzj5j7nhclpmxu46bnr3ymq`. When this request reaches the IndexStore, it queries the Singularity database on the `car_blocks` table, getting the following record:

Author: ```suggestion
Once the client verifies and decodes the block, it MAY intend to traverse the DAG and request following blocks (DAG links) from the server. For instance, a follow-up request MAY be performed to get a link with CID `bafkreihxpvy6y7aloo5s4entwbnkhaqzczgwzj5j7nhclpmxu46bnr3ymq`. When this request reaches the IndexStore, it queries the Singularity database on the `car_blocks` table, getting the following record:
```

  // length of the data
  "length": 57523,
  // offset of the data in the location byte stream
  "offset": 2097152
}
```

### More complex response pattern

In this pattern, we consider a more complex response where all the blocks that are part of the DAG behind a given CID are streamed right away. Let's consider the same CID as in the previous case: `bafybeidxxkuao2zamg5rd7pypqrhrjmaqayxp7wr5ojmqdqbtpvzje74au`.

Once the IndexStore backed by the Singularity database receives this request to find all the blocks that have some relationship to create the DAG with CID `bafybeidxxkuao2zamg5rd7pypqrhrjmaqayxp7wr5ojmqdqbtpvzje74au`, it can query the `car_blocks` table as previously, obtaining the exact same record: 

```json
{
  "id": 377351956,
  "cid": "bafybeidxxkuao2zamg5rd7pypqrhrjmaqayxp7wr5ojmqdqbtpvzje74au",
  "carOffset": 2154851,
  "carBlockLength": 197,
  "varint": "wwE=",
  "rawBlock": "EiwKJAFVEiD3fXHsfAtzuy4Rs7Bao4IZFk1sp6n7TiW9l6c8Fsd4ZBIAGICAQBIsCiQBVRIgknipVUA6k+D7/at6kJIKw3djt7YFeTtVtexkvTN3gc8SABiAgEASLAokAVUSIGLF/7OZ+6M6WJKEtwumIV6ucqv+G0o2Z3ZQ1CCgJ1ekEgAYs8EDChMIAhizwYMBIICAQCCAgEAgs8ED",
  "fileOffset": 0,
  "carId": 17482,
  "fileId": null
}
```

> The reader can already yield this like previously as:

Author: ```suggestion
The reader can already yield the index record as previously:
```

  "offset": 0,
	// associated records
	"subRecords": []
}
```

However, instead of stopping as in the previous `request-response pattern`, it will try to find further records. For this, it will query the `files` table now with the exact same CID, joining the results with the `car_blocks` entries on the `file_id` of the file found, as well as the `cars` and `storages` tables, resulting in the following records:

- File record (separated for ease of illustration):

```json
{
	"id": 2085318,
	"cid": "bafybeidxxkuao2zamg5rd7pypqrhrjmaqayxp7wr5ojmqdqbtpvzje74au",
	"path": "001-Al-Fatihah.mp3",
	"hash": "03f41c1eb44caa4f8e55768e6b2fc127",
	"size": 2154675,
	"lastModifiedNano": 1746799656000000000,
	"attachmentId": 590,
	"directoryId": 18042
}
```

- Storage:

```json
{
	"id": 400,
	"name": "foo",
	"createdAt": "2024-12-13T17:37:44.703659Z",
	"updatedAt": "2024-12-13T17:37:44.703659Z",
	"type": "example.com",
	"path": "foo",
	"config": {
		"disable_checksum": "true",
		"encoding": "Slash,LtGt,CrLf,Del,Ctl,InvalidUtf8,Dot",
		"endpoint": "https://s3.us.example.com",
		"front_endpoint": "https://example.com",
		"wait_archive": "0s"
	},
	"clientConfig": {}
},
```

> - Blocks:

Author: ```suggestion
- Blocks with `fileId`:
```

		"multihash": "MH(bafkreihxpvy6y7aloo5s4entwbnkhaqzczgwzj5j7nhclpmxu46bnr3ymq)",
		// Type of the record
		"type": "BLOB",
		// hash digest of the location or Path
		"location": "https://example.com/download/foo/001-Al-Fatihah.mp3",
		// length of the data
		"length": 1048576,
		// offset of the data in the location byte stream
		"offset": 0
	},
	{
		// MultihashDigest identifying the record
		"multihash": "MH(bafkreiespcuvkqb2spqpx7nlpkijecwdo5r3pnqfpe5vlnpmms6tg54bz4)",
		// Type of the record
		"type": "BLOB",
		// hash digest of the location or Path
		"location": "https://example.com/download/foo/001-Al-Fatihah.mp3",
		// length of the data
		"length": 1048576,
		// offset of the data in the location byte stream
		"offset": 1048576
	},
	{
		// MultihashDigest identifying the record
		"multihash": "MH(bafkreidcyx73hgp3um5freuew4f2mik6vzzkx7q3ji3go5sq2qqkaj2xuq)",
		// Type of the record
		"type": "BLOB",
		// hash digest of the location or Path
		"location": "https://example.com/download/foo/001-Al-Fatihah.mp3",
		// length of the data
		"length": 57523,
		// offset of the data in the location byte stream
		"offset": 2097152
	}
]
```

## Relevant details

### Database Queries

> These SQL queries are used to retrieve index records for a given CID from a **Singularity-like SQLite database schema**. They serve as the basis for constructing byte range references to content stored simulating CAR files within object storage.

Author: ```suggestion
These SQL queries are used to retrieve index records for a given CID from a **Singularity-like SQLite database schema**. They serve as the basis for constructing byte range references to content stored within object storage.
```
